test/extensions/filters/network/common/fuzz/network_readfilter_fuzz_test.cc

```diff
@@ -9,41 +9,86 @@
 #include "test/fuzz/fuzz_runner.h"
 #include "test/test_common/test_runtime.h"
 
+#include "libprotobuf_mutator/src/libfuzzer/libfuzzer_macro.h"
+#include "src/text_format.h" // from libprotobuf_mutator/
+
 namespace Envoy {
 namespace Extensions {
 namespace NetworkFilters {
 
-DEFINE_PROTO_FUZZER(const test::extensions::filters::network::FilterFuzzTestCase& input) {
+void ensuredValidFilter(unsigned int seed, envoy::config::listener::v3::Filter* config) {
+  // TODO(jianwendong): After extending to cover all the filters, we can use
+  // `Registry::FactoryRegistry<
+  // Server::Configuration::NamedNetworkFilterConfigFactory>::registeredNames()`
+  // to get all the filter names instead of calling `UberFilterFuzzer::filter_names()`.
+  static const auto filter_names = UberFilterFuzzer::filterNames();
+  static const auto factories = Registry::FactoryRegistry<
+      Server::Configuration::NamedNetworkFilterConfigFactory>::factories();
+
+  // Choose a valid filter name.
+  if (std::find(filter_names.begin(), filter_names.end(), config->name()) ==
+      std::end(filter_names)) {
+    absl::string_view filter_name = filter_names[seed % filter_names.size()];
+    if (filter_name != config->name()) {
+      // Clear old config, or unpacking non-suitable value may crash.
+      config->clear_typed_config();
+      config->set_name(std::string(filter_name));
+    }
+  }
+  // Set the corresponding type_url for Any.
+  auto& factory = factories.at(config->name());
+  config->mutable_typed_config()->set_type_url(absl::StrCat(
+      "type.googleapis.com/", factory->createEmptyConfigProto()->GetDescriptor()->full_name()));
+}
+
+static void TestOneProtoInput(const test::extensions::filters::network::FilterFuzzTestCase&);
+using FuzzerProtoType = test::extensions::filters::network::FilterFuzzTestCase;
+
+extern "C" size_t LLVMFuzzerCustomMutator(uint8_t* data, size_t size, size_t max_size,
+                                          unsigned int seed) {
+  // mutate the config part of the fuzzer only with a probability of
+  static const unsigned config_mutation_probability = 1 /* / 100 */;
+  static protobuf_mutator::Mutator mutator = [seed] {
+    protobuf_mutator::Mutator mutator;
+    mutator.Seed(seed);
+    return mutator;
+  }();
+  static ProtobufMessage::ValidatedInputGenerator generator(
+      seed, ProtobufMessage::composeFiltersAnyMap(), 20);
+  static Random::PsuedoRandomGenerator64 random;
+  static bool _random_inited = [seed] {
+    random.initializeSeed(seed);
+    return true;
+  }();
+  _random_inited = !_random_inited;
+
+  FuzzerProtoType input;
+  input.ParseFromString({reinterpret_cast<const char*>(data), size});
+  if (random.random() % 100 < config_mutation_probability) {
+    test::extensions::filters::network::FuzzHelperForActions actions;
+    *actions.mutable_actions() = *input.mutable_actions();
+    mutator.Mutate(&actions, max_size);
```

Comment:
I think there's a missing copying back the actions to the input's actions.

Taking a step back, shouldn't the protobuf message itself be changed, like you did before?

Reply:
I agree, it's safer and better to understand to copy the actions back into the input. I had the impression, that due the pointer nature of both members they were just hooking pointers around.

You really liked the way before? I only went that way, because of the interface of `mutator.Mutate()` not being able to take a repeated field. So the clearest way would be to extend `Mutate()`, but that means to modify libprotobuf_mutator, which take a lot of time. Yes, I am also not happy about copying all this data, but we already copy so much, that it does not matter anymore.

Reply:
I agree that the way before isn't ideal, but I'm not sure that pointers are used, because the indirection will probably copy bits over.
Maybe `std::move()` can be used in this case...

Reply:
`std::move()` sounds good. Applied.